pil-analyzer/src/type_inference.rs

```diff
@@ -688,16 +705,60 @@ impl<'a> TypeChecker<'a> {
                 _ => {}
             };
         }
-        let inferred_type = self.infer_type_of_expression(expr)?;
-        self.unifier
-            .unify_types(inferred_type.clone(), expected_type.clone())
-            .map_err(|err| {
-                format!(
-                    "Error checking sub-expression {expr}:\nExpected type: {}\nInferred type: {}\n{err}",
-                    self.type_into_substituted(expected_type.clone()),
-                    self.type_into_substituted(inferred_type)
-                )
-            })
+
+        let inferred_type = self.infer_type_of_expression(expr, kind)?;
+        self.unify_based_on_kind(expr, kind, &inferred_type, expected_type)
+    }
+
+    fn unify_based_on_kind(&mut self, expr: &Expression, kind: Option<FunctionKind>, inferred_type: &Type, expected_type: &Type) -> Result<(), String> {
+        match kind {
+            Some(FunctionKind::Constr) => {
+                self.unifier
+                    .unify_types(inferred_type.clone(), expected_type.clone())
+                    .map_err(|err| {
+                        format!(
+                            "Error checking expression {expr}:\nExpected type: {}\nInferred type: {}\n{err}",
+                            self.type_into_substituted(expected_type.clone()),
+                            self.type_into_substituted(inferred_type.clone())
+                        )
+                    })
+            }
+            _ => {
+                if !Self::contains_constr(inferred_type) {
+                    self.unifier
+                        .unify_types(inferred_type.clone(), expected_type.clone())
+                        .map_err(|err| {
+                            format!(
+                                "Error checking expression {expr}:\nExpected type: {}\nInferred type: {}\n{err}",
+                                self.type_into_substituted(expected_type.clone()),
+                                self.type_into_substituted(inferred_type.clone())
+                            )
+                        })
+                } else {
+                    Err(format!(
+                        "Error Constr type inside Pure/Query function {expr}:\nExpected type: {}\nInferred type: {}\n",
```

Comment:
I’m not sure I understand this correctly, but the type checker should not care about the Constr type. It is totally fine to use that type in any function. One thing we need to make sure is that is that the type of a statement can only be the empty tuple, except inside functions of kind constr, where it is also allowed to have the type passed in from the pil analyser (which is Constr, but the type inference does not know of care about the concrete type)